### PR TITLE
Add saving of logs to disk & combining into CSV

### PR DESCRIPTION
**Pull request by michaeltremeer**
The code adds the following:
- Adds a `json-save-dir` arg to the load parser, so that JSON logs can be saved to disk
    - A filehandler is added to the base logger, and the output folders & file created automatically.
    - Pipes the logs out, including most of the run args (so that each log file contains all relevant configuration when being compared later).

- Adds a `combine_logs` subcommand to the parser
    - Extracts run information and final metrics (prior to termination/max requests/duration) from raw log files, and combines them into a single output CSV file, ready for analysis.

### CODE REVIEW

File: benchmark/statsaggregator.py [hunk 1]
```diff
@@ -153,9 +158,9 @@ def _dump(self):
                   "95th": util_95th,
                },
             }
-            print(json.dumps(j), flush=True)
+            logger.info(json.dumps(j))
          else:
-            print(f"{timestamp} rpm: {rpm:<5} requests: {self.requests_count:<5} failures: {self.failed_count:<4} throttled: {self.throttled_count:<4} tpm: {tokens_per_minute:<6} ttft_avg: {ttft_avg:<6} ttft_95th: {ttft_95th:<6} tbt_avg: {tbt_avg:<6} tbt_95th: {tbt_95th:<6} e2e_avg: {e2e_latency_avg:<6} e2e_95th: {e2e_latency_95th:<6} util_avg: {util_avg:<6} util_95th: {util_95th:<6}", flush=True)
+            logger.info(f"rpm: {rpm:<5} requests: {self.requests_count:<5} failures: {self.failed_count:<4} throttled: {self.throttled_count:<4} tpm: {tokens_per_minute:<6} ttft_avg: {ttft_avg:<6} ttft_95th: {ttft_95th:<6} tbt_avg: {tbt_avg:<6} tbt_95th: {tbt_95th:<6} e2e_avg: {e2e_latency_avg:<6} e2e_95th: {e2e_latency_95th:<6} util_avg: {util_avg:<6} util_95th: {util_95th:<6}")
```
Comment (technicianted) [inline benchmark/statsaggregator.py:163]: The idea behind using `print` and not `logger` is to redirect to stdout vs stderr such that you can use shell redirection to only get the stats or jsonl. You will need to add conditionals here to make sure that this behvior doesn't break.

File: benchmark/jsonloganalysis.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import argparse
```
Comment (technicianted) [inline benchmark/jsonloganalysis.py:1]: I think this feature is too specific for this tool. In addition, but outputting stats in jsonl, you can use simple command line tools such as `jq` to aggregate.

Perhaps you should take this feature out in a separate PR for better discussion.

File: benchmark/bench.py [hunk 3]
```diff
@@ -45,7 +50,28 @@ def main():
...
 
+    combine_logs_parser = sub_parsers.add_parser("combine_logs", help="Combine JSON logs from previous runs into a single CSV.")
+    combine_logs_parser.add_argument("logdir", type=str, help="Directory containing the log files.")
+    combine_logs_parser.add_argument("savepath", type=str, help="Path to save the output output CSV.")
+    combine_logs_parser.add_argument("--load-recursive", action="store_true", help="Whether to load logs in all subdirectories of log_dir.")
+    combine_logs_parser.set_defaults(func=combine_logs_to_csv)
+
     args = parser.parse_args()
+
+    if args.func is load and args.log_save_dir is not None:
```
Comment (technicianted) [inline benchmark/bench.py:61]: I'm wondering how is this different from using shell redirection of stderr? I could do something like this:
```
$ python -m benchmark.bench load ... 2> my/output/dir/output.log
```
because this is only stderr, I will only get the `logger` in `output.log`

File: README.md [hunk 4]
```diff
@@ -33,7 +33,7 @@ $ docker run azure-openai-benchmarking load --help
...
+1. **Run your test long enough to reach a stable state**. Throttling is based on the total compute you have deployed and are utilizing. The utilization includes active calls. As a result you will see a higher call rate when ramping up on an unloaded deployment because there are no existing active calls being processed. Once your deplyoment is fully loaded with a utilzation near 100%, throttling will increase as calls can only be processed as earlier ones are completed. To ensure an accurate measure, set the duration long enough for the throughput to stabilize, especialy when running at or close to 100% utilization. Also note that once the test ends (either by termination, or reaching the maximum duration or number of requests), any pending requests will continue to drain, which can result in lower throughput values as the load on the endpoint gradually decreases to 0.
```
Comment (technicianted) [inline README.md:36]: Great clarification!

**Review by technicianted — CHANGES_REQUESTED** [review]
Thanks for your contribution!

**Comment by michaeltremeer** [comment]
Good points in the comments, and this is a stretch for the core functionality of the repo - agreed. The only note I would say though (inspired by my own use): With shell redirection, you do miss any warnings or errors that occurred during the run, as well as any run parameters which make it easier to tell what the config was for each run (useful when you're analysing a number of permutations of shape, model and concurrency). You also miss the time the run finished or was terminated (which means the post-test period where requests are still draining can impact the numbers). 

If you think it's valid and worth a change, I can resubmit just the first set of changes for saving a single run's logs, otherwise I can just submit the updated README for clarity.

**Comment by technicianted** [comment]
>> you do miss any warnings or errors that occurred during the run
You can redirect both stdout and stderr to the same file if needed:
```
$ python -m benchmark.bench load ... > my/output/dir/full_output.log 2>&1
```
or separate files:
```
$ python -m benchmark.bench load ... > my/output/dir/stats_output.log 2> my/output/dir/logs.log
```
Does that cover the use-case?

>> as well as any run parameters which make it easier to tell what the config was for each run 
Yes I agree this is super useful. Let's add that in a separate PR. 

>> I can resubmit just the first set of changes for saving a single run's logs, otherwise I can just submit the updated README for clarity.
Yes let's break them down. Generally many smaller PRs is always better :)

Thanks!
